compute/cker/include/cker/train/operation/AvgPool.h

```diff
@@ -0,0 +1,150 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __NNFW_CKER_TRAIN_OPERATION_AVGPOOL_H__
+#define __NNFW_CKER_TRAIN_OPERATION_AVGPOOL_H__
+
+#include "cker/Shape.h"
+#include "cker/Utils.h"
+#include "cker/eigen/Utils.h"
+
+#include <Eigen/Core>
+
+namespace nnfw
+{
+namespace cker
+{
+namespace train
+{
+inline void AvgPool2D(const PoolParams &params, const Shape &input_shape, const float *input_data,
+                      const Shape &output_shape, float *output_data)
+{
+  assert(input_shape.DimensionsCount() == 4);
+  assert(output_shape.DimensionsCount() == 4);
+  const int batches = MatchingDim(input_shape, 0, output_shape, 0);
+  const int input_height = input_shape.Dims(1);
+  const int input_width = input_shape.Dims(2);
+  const int output_height = output_shape.Dims(1);
+  const int output_width = output_shape.Dims(2);
+  const int stride_height = params.stride_height;
+  const int stride_width = params.stride_width;
+
+  // TODO(benoitjacob) make this a proper reference impl without Eigen!
+  const auto in_mat = MapAsMatrixWithLastDimAsRows(input_data, input_shape);
+  auto out_mat = MapAsMatrixWithLastDimAsRows(output_data, output_shape);
+
+  // Prefill the output to 0.
+  out_mat.setZero();
+
+  for (int b = 0; b < batches; ++b)
+  {
+    for (int h = 0; h < output_height; ++h)
+    {
+      for (int w = 0; w < output_width; ++w)
+      {
+        // (h_start, h_end) * (w_start, w_end) is input range
+        // that output is projected from.
+        int h_start = h * stride_height - params.padding_values.height;
+        int h_end = std::min(h_start + params.filter_height, input_height);
+        h_start = h_start < 0 ? 0 : h_start;
+
+        int w_start = w * stride_width - params.padding_values.width;
+        int w_end = std::min(w_start + params.filter_width, input_width);
+        w_start = w_start < 0 ? 0 : w_start;
+
+        int count = (h_end - h_start) * (w_end - w_start);
+        if (h_end <= 0 || w_end <= 0 || count <= 0 || h_start >= input_height ||
+            w_start >= input_width)
+          continue;
+
+        int out_offset = NodeOffset(b, h, w, output_height, output_width);
+        for (int ph = h_start; ph < h_end; ++ph)
+        {
+          for (int pw = w_start; pw < w_end; ++pw)
+          {
+            int in_offset = NodeOffset(b, ph, pw, input_height, input_width);
+            out_mat.col(out_offset) += in_mat.col(in_offset);
+          }
+        }
+        out_mat.col(out_offset) /= count;
+      }
+    }
+  }
+
+  out_mat.cwiseMin(params.float_activation_min).cwiseMax(params.float_activation_max);
+}
```

Comment:
"cker" has already had inference forwarding kernel in `cker/operation/AveragePool.h`. Is this function neccesary?

Reply:
> "cker" has already had the forwarding function in `cker/operation/AveragePool.h`. Is this function neccesary?

I thought train needs its own forward function. (ex. `nnfw::cker::train::MaxPool2D`)

https://github.com/Samsung/ONE/blob/0b7b37ab4f4904b7ecdf7ec9bd09e0ef66be9402/runtime/onert/backend/train/ops/PoolLayer.cc#L78-L87

If `nnfw::cker::AveragePool()` is callable for forward function in `train/ops/PoolLayer.cc` and `nnfw::cker::train::AvgPool2DGrad` for backward, I think this function is unnecessary.

Reply:
> Updating like this looks good for me :)

I've changed `nnfw::cker::Averagepool()`,but found test fails.
All `calculated_output` filled with `0.0f`.

But I don't know the difference between 2 codes.

https://github.com/Samsung/ONE/blob/815511969fca1a7276ad99fa199d5ebcc202e41a/compute/cker/include/cker/train/operation/MaxPool.h#L126

https://github.com/Samsung/ONE/blob/815511969fca1a7276ad99fa199d5ebcc202e41a/compute/cker/include/cker/operation/AveragePool.h#L98-L103

Reply:
> Isn't two codes same? 

Probably yes. In Eigen, "cwise" means "coefficient-wise", and the term "coefficient-wise" means that when performing matrix operations, the operation is performed individually for each element.

Reply:
> Probably yes. In Eigen, "cwise" means "coefficient-wise", and the term "coefficient-wise" means that when performing matrix operations, the operation is performed individually for each element.

So, the code `out_mat.cwiseMin(params.float_activation_min)` means
```c++
 const int flat_size = output_shape.FlatSize(); 
 for (int i = 0; i < flat_size; ++i) 
 { 
   output_data[i] = std::min(output_data[i], params.float_activation_min);
 } 
```

and `.cwiseMax(params.float_activation_max) ` means

```c++
 const int flat_size = output_shape.FlatSize(); 
 for (int i = 0; i < flat_size; ++i) 
 { 
   output_data[i] = std::max(output_data[i], params.float_activation_max);
 } 
```


Doesn't it make all `output_data` to `params.float_activation_max`?

Reply:
@zetwhite Could you explain the difference?

Reply:
> Doesn't it make all output_data to params.float_activation_max?

You are right. They are should be exchanged.

Reply:
> @zetwhite Could you explain the difference? 

Ah, Thank you for notice :+1: It might be my mistake. 
I'll change the code like this : 
```cpp 
// train/maxpool.h 
 out_mat.cwiseMin(params.float_activation_max).cwiseMax(params.float_activation_min);
```